Include/internal/pycore_critical_section.h

```diff
@@ -0,0 +1,204 @@
+#ifndef Py_INTERNAL_CRITICAL_SECTION_H
+#define Py_INTERNAL_CRITICAL_SECTION_H
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_lock.h"        // PyMutex
+#include "pycore_pystate.h"     // _PyThreadState_GET()
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// Implementation of Python critical sections: helpers to replace the global
+// interpreter lock with fine grained locking.
+//
+// A Python critical section is a region of code that can only be executed by
+// a single thread at at time. The regions begin with a call to
+// Py_BEGIN_CRITICAL_SECTION and ends with Py_END_CRITICAL_SECTION. Active
+// critical sections are *implicitly* suspended whenever the thread calls
+// `_PyThreadState_Detach()` (i.e., at tiems when the interpreter would have
+// released the GIL).
+//
+// The most recent critical section is resumed when `_PyThreadState_Attach()`
+// is called. See `_PyCriticalSection_Resume()`.
+//
+// The purpose of implicitly ending critical sections is to avoid deadlocks
+// when locking multiple objects. Any time a thread would have released the
```

Comment:
I've rewritten this using more excerpts from PEP 703. https://peps.python.org/pep-0703/#python-critical-sections. Deadlocks avoidance is the primary design goal of this API.

Please let me know if the updated docs make more sense and what additional clarifications would be beneficial.

> How does attach/detach relate to locking multiple objects?

Threads that lock the same objects in a different order would deadlock. This could happen from nested critical sections [^1]. We could avoid it by suspending active critical sections any time we begin a new critical section, so only one critical section is active per thread. That would work, but for performance reasons we do a slight modification: we only suspend critical sections in `_PyThreadState_Detach()`. That's faster in the common case (fewer lock/unlocks) and still avoids deadlock because any thread that would deadlock calls `_PyThreadState_Detach()` (which in turn suspends outer critical sections, avoiding the deadlock).

[^1]: It could also happen if a single critical section  acquires multiple locks (like `Py_BEGIN_CRITICAL_SECTION2`). That's why `Py_BEGIN_CRITICAL_SECTION2` sorts the locks so any call to `Py_BEGIN_CRITICAL_SECTION2` with the same objects always locks them in the same order.

Reply:
Thanks for explaining.